src/katsdpmodels/fetch.py

```diff
@@ -0,0 +1,359 @@
+################################################################################
+# Copyright (c) 2020, National Research Foundation (SARAO)
+#
+# Licensed under the BSD 3-Clause License (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+#   https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Fetch models over HTTP."""
+
+import contextlib
+import errno
+import io
+import hashlib
+import re
+import logging
+import os
+import urllib.parse
+from typing import Dict, Tuple, Optional, Type, TypeVar, Any, cast
+from typing_extensions import Protocol
+
+import requests
+import requests_file
+
+from . import models
+
+
+MAX_ALIASES = 30
+_logger = logging.getLogger(__name__)
+_F = TypeVar('_F', bound='Fetcher')
+_M = TypeVar('_M', bound='models.Model')
+
+
+class Session(Protocol):
+    """Generalization of :class:`requests.Session`."""
+
+    def get(self, url: str, **kwargs) -> requests.Response: ...   # pragma: nocover
+    def head(self, url: str, **kwargs) -> requests.Response: ...  # pragma: nocover
+    def close(self) -> None: ...                                  # pragma: nocover
+
+
+class HttpFile(io.RawIOBase):
+    """File-like object that fetches byte ranges via HTTP.
+
+    This requires the server to advertise for support byte-range requests and
+    to provide a Content-Length. It is currently *not* robust against the
+    content changing.
+
+    Raises
+    ------
+    FileNotFoundError
```

Comment:
Any chance of some out of range issue occurring with a request ?

Reply:
I guess if the server truncates the file under us? But I haven't tried to catch HTTP errors in the read functions, because RequestException inherits from IOError anyway so code that thinks this is a file and catches IOError will catch those exceptions anyway.

Reply:
Fair enough.